DM-26214: Switch to using sourceTable_visit parquet files instead of src FITS files

bechtol opened the pull request:
This PR adds base classes that use `sourceTable_visit` parquet files as input to per-visit and per-detector metrics.

To test, run a command like

`pipetask --long-log run -b /repo/main/butler.yaml --register-dataset-types -p testpipe.yaml -d "visit=35892 AND skymap='hsc_rings_v1' AND instrument='HSC'" --output u/kbechtol/sourcetable_test -i HSC/runs/RC2/w_2021_18/DM-29973 --timeout 999999`

With testpipe.yaml pipeline file something like

```
description: Compute metrics from sourceTable_visit catalogs
tasks:
  nsrcMeasVisitTable:
    class: lsst.faro.measurement.VisitTableMeasurementTask
    config:
      connections.package: info
      connections.metric: nsrcMeasVisitTable
      python: |
        from lsst.faro.base import NumSourcesTask
        config.measure.retarget(NumSourcesTask)
        config.columns = 'coord_ra, coord_dec, visit'

description: Compute per-detector metrics from sourceTable_visit catalogs
tasks:
  nsrcMeasDetectorTable:
    class: lsst.faro.measurement.DetectorTableMeasurementTask
    config:
      connections.package: info
      connections.metric: nsrcMeasDetectorTable
      python: |
        from lsst.faro.base import NumSourcesTask
        config.measure.retarget(NumSourcesTask)
        config.columns = 'coord_ra, coord_dec, visit, detector'
```

In the current PR version, there are new base classes that use `sourceTable_visit` rather than replacing the previous base classes that use `src` catalogs, i.e., we retain both versions for now. Perhaps we should discuss at what point / if we want to completely replace the versions that use `src` catalogs. Another question for this PR is whether we want to add base classes for the object tables and for creating the matching catalogs, or whether we want those as separate PRs.

Currently, this PR does not have new/modified pipeline yaml files included.

bechtol commented:
Added a commit to demonstrate access to reference catalogs in `DetectorMeasurement.py`.  In this version, one can choose between Gaia and PS1, and can optionally apply proper motions or colorterms, respectively.  

To test, one can use a simple pipeline similar to the following

```
description: Compute per-detector metrics from sourceTable_visit catalogs
tasks:
  nsrcMeasDetectorTable:
    class: lsst.faro.measurement.DetectorTableMeasurementTask
    config:
      connections.package: info
      connections.metric: nsrcMeasDetectorTable
      #connections.refDataset: 'gaia_dr2_20200414'
      connections.refDataset: 'ps1_pv3_3pi_20170110'
      python: |
        from lsst.faro.base import NumSourcesTask
        config.measure.retarget(NumSourcesTask)
        config.columns = 'coord_ra, coord_dec, visit, detector'
```

and call pipetask from the command line to run for a single detector as follows

`pipetask --long-log run -b /repo/main/butler.yaml --register-dataset-types -p testpipe.yaml -d "visit=35892 AND detector=14 AND skymap='hsc_rings_v1' AND instrument='HSC'" --output u/kbechtol/sourcetable_test -i HSC/runs/RC2/w_2021_18/DM-29973 --timeout 999999`

bechtol commented:
Based on our discussions, recent commits put the base classes that use parquet tables into separate files so that they are more readily distinguished from the versions of base classes that do not use parquet table inputs.

Added pipeline files for the per-visit and per-detector metrics.

We have decided not to add the base classes that use parquet tables for all analysis contexts in this PR, but rather have the per-visit and per-detector examples and we'll add versions for other analysis contexts in separate PRs.

parejkoj commented:
Please fix the PR title to match the dev guide: `DM-26214: Switch ...`
https://developer.lsst.io/work/flow.html#make-a-pull-request

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+            center = lsst.geom.SpherePoint(butlerQC.quantum.dataId.region.getBoundingCircle().getCenter())
+            radius = butlerQC.quantum.dataId.region.getBoundingCircle().getOpeningAngle()
+
+            epoch = butlerQC.quantum.dataId.records['visit'].timespan.begin
```
Is this an astropy.date?

bechtol replied:
`type(epoch)` returns `<class 'astropy.time.core.Time'>`

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+            #     refCat.extend(handle.get())
+            # if not refCat.isContiguous():
+            #     refCat = refCat.copy(deep=True)
+            # inputs['refcat'] = refCat
```
Please remove commented-out lines.

bechtol replied:
Done

parejkoj commented on python/lsst/faro/measurement/VisitTableMeasurement.py:
```diff
@@ -0,0 +1,48 @@
+
+class VisitTableMeasurementTaskConfig(CatalogMeasurementBaseTaskConfig,
+                                      pipelineConnections=VisitTableMeasurementTaskConnections):
+    columns = pexConfig.Field(doc="Columns from sourceTable_visit to load.",
```
`ListField` here as well.

bechtol replied:
Done

parejkoj reviewed (changes requested):
Do you have any tests of this code? I have doubts that the refcat loading part will work as written.

Quite a few comments about code style and organization, and some concerns about how the runQuantum/run methods interact with the loaders.

erykoff commented:
@bechtol FYI, the PR should have the title `DM-26214: Switch to using sourceTable_visit parquet files instead of src FITS files`, as per DM dev guide: https://developer.lsst.io/work/flow.html#make-a-pull-request

erykoff commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+            if self.config.connections.refDataset == 'gaia_dr2_20200414':
+                # Apply proper motions for Gaia catalog
+                config.refObjLoader.requireProperMotion = True
+                config.refObjLoader.anyFilterMapsToThis = 'phot_g_mean'
```
To be clear, the `LoadReferenceCatalogTask` uses the `ReferenceObjectLoader` and also adds on color corrections and object selection.  It does support `anyFilterMapsToThis` and also applies proper motions if an epoch is set, see: https://github.com/lsst/pipe_tasks/blob/master/python/lsst/pipe/tasks/loadReferenceCatalog.py#L174-L202

parejkoj replied:
Thanks for the clarification, @erykoff !

SimonKrughoff commented on pipelines/measurement/measurement_detector_table.yaml:
```diff
@@ -0,0 +1,12 @@
+      connections.package: info
+      connections.metric: nsrcMeasDetectorTable
+      #connections.refDataset: 'gaia_dr2_20200414'
+      connections.refDataset: 'ps1_pv3_3pi_20170110'
```
Rather than have a commented connection, if the goal here is to document which catalogs are available, I'd much rather they be listed in a comment rather than have the options be commented in valid code.  I.e. if it's there for purely informational purposes, it should be in a comment to make that clear.

bechtol replied:
Made Gaia the default. Listing Gaia and PS1 as a separate comment now instead of commented options.

SimonKrughoff commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+                                             deferLoad=True)
+
+    refcat = pipeBase.connectionTypes.PrerequisiteInput(
+        doc="Reference catalog.",
```
I think I'd prefer the indentation for this attribute match the others in this file for readability

bechtol replied:
Making consistent.

SimonKrughoff commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+        multiple=True
+    )
+
+    measurement = pipeBase.connectionTypes.Output(doc="Per-detector measurement.",
```
If John prefers the other format, we can go that way.  I just want consistency.  For the record we use this format in the other files, I thnk

bechtol replied:
I do prefer the format with newline and all single-indent. Agree with consistency. We will probably want to update all the class names for connections and configs to follow DM convention, so that might be a time to make this all consistent throughout the package.

SimonKrughoff commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+                config.refObjLoader.load(os.path.join(getPackageDir('obs_subaru'),
+                                                      'config',
+                                                      'filterMap.py'))
+                config.colorterms.load(os.path.join(getPackageDir('obs_subaru'),
```
I think this override should probably go in a subaru specific pipeline task

bechtol replied:
This is now removed. 

I have confirmed that I can load a ColorTermLibrary in a pipeline yaml file by adding lines like
```
        import os
        config.refObjLoader.refObjLoader.load(os.path.join(getPackageDir('obs_subaru'),
                                                           'config',
                                                           'filterMap.py'))
        config.refObjLoader.colorterms.load(os.path.join(getPackageDir('obs_subaru'),
                                                         'config',
                                                         'colorterms.py'))
```
However, I'm not sure if this is the "standard" way of passing in the ColorTermLibrary

parejkoj replied:
Yes, that is the standard approach. Eventually, once we've settled on the RFC-775 configuration layout, that config would live in drp_pipe (probably?) in an instrument-specific override file.

SimonKrughoff commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+    _DefaultName = "detectorTableMeasurementTask"
+
+    def run(self, catalog, dataIds, refcat, refcatCalib):
+        return self.measure.run(catalog, self.config.connections.metric, dataIds)
```
I think it's so we can print log information as we process the quanta.

SimonKrughoff replied:
It also doesn't pass refcatCalib.  I think it should probably pass both of those things on.

bechtol replied:
Agree that we definitely want to pass `refcat` and `refcatCalib` in the `run` method. I think these should be keyword arguments because not all measurement tasks will need them.

bechtol replied:
I _think_ the `dataIds` will not be needed. In other `faro` analysis contexts, the dataIds have been used to determine what bands are being analyzed, e.g., to compare with stellar locus (`WPerpTask` in `python/lsst/faro/measurement/TractMeasurementTasks.py`).

parejkoj replied:
If a task needs to determine the filter bands, you should pass in a `FilterLabel` or `list` of `FilterLabel`s.

bechtol replied:
Thanks. That seems a cleaner approach. I've updated the following ticket to reflect this guidance.

https://jira.lsstcorp.org/browse/DM-31061

SimonKrughoff commented on python/lsst/faro/measurement/VisitTableMeasurement.py:
```diff
@@ -0,0 +1,48 @@
+
+    def runQuantum(self, butlerQC, inputRefs, outputRefs):
+        inputs = butlerQC.get(inputRefs)
+        columns = [_.strip() for _ in self.config.columns.split(',')]
```
As John mentioned, this manipulation isn't necessary if you use a `ListField` in the config

bechtol replied:
Done

SimonKrughoff commented on python/lsst/faro/measurement/VisitTableMeasurement.py:
```diff
@@ -0,0 +1,48 @@
+    def runQuantum(self, butlerQC, inputRefs, outputRefs):
+        inputs = butlerQC.get(inputRefs)
+        columns = [_.strip() for _ in self.config.columns.split(',')]
+        inputs['catalog'] = inputs['catalog'].get(parameters={'columns': columns})
```
I'm a little confused by this.  I think `butlerQC.get` should have returned a `pandas.DataFrame` already, and the get method on a `DataFrame` does not take a parameters argument.  Did you perhaps mean to pop off the catalog entry from `inputRefs` and use that to get the catalog and put it in `inputs['catalog']`?

bechtol replied:
The objective here was to selectively load a subset of columns from the parquet table. The input for the `sourceTable_visit` uses a deferred load so that one can use `get` her with the `parameters` argument specified. Is there another way to do this?

erykoff replied:
The problem is that you are replacing a catalog reference with a data frame. You should use different variable names.

bechtol replied:
Now using a different variable name, `catalog` instead of `inputs['catalog']` so that `inputs['catalog']` is not modified.

SimonKrughoff reviewed (approved):
I do have concerns that I don't think the code would not run as intended as written which points to a lack of tests.   We have taken the approach that testing the algorithmic code is more important than testing the task code, but I do think we need to think about how we can try to get better coverage.  I'm happy making that another PR, but we should move that up in priority.

I'm marking this approved so as not to hold things up while I'm on vacation, but do consider other reviewers comments.

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+import os
```
Thanks. Done.

bechtol replied:
Created a separate ticket to do this consistently in faro:
https://jira.lsstcorp.org/browse/DM-31135

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+
+class DetectorTableMeasurementTaskConfig(CatalogMeasurementBaseTaskConfig,
+                                         pipelineConnections=DetectorTableMeasurementTaskConnections):
+    columns = pexConfig.Field(doc="Columns from sourceTable_visit to load.",
```
Thank you. I didn't realize `ListField` existed.

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+                                         pipelineConnections=DetectorTableMeasurementTaskConnections):
+    columns = pexConfig.Field(doc="Columns from sourceTable_visit to load.",
+                              dtype=str, default='coord_ra, coord_dec, detector')
+    refDataset = pexConfig.Field(doc="Reference dataset to use.",
```
Good catch. I thought maybe it was needed, but have confirmed that it is not needed.

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+                                 dtype=str, default='')
+
+
+class DetectorTableMeasurementTask(CatalogMeasurementBaseTask):
```
Done

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+        catalog = inputs['catalog'].get(parameters={'columns': columns})
+        selection = (catalog['detector'] == butlerQC.quantum.dataId['detector'])
+        inputs['catalog'] = catalog[selection]
+        inputs['dataIds'] = [butlerQC.registry.expandDataId(inputRefs.catalog.datasetRef.dataId)]
```
This was adding a new `dataIds` field

bechtol commented on python/lsst/faro/measurement/VisitTableMeasurement.py:
```diff
@@ -0,0 +1,48 @@
+                              dtype=str, default='coord_ra, coord_dec')
+
+
+class VisitTableMeasurementTask(CatalogMeasurementBaseTask):
```
Added brief docstring

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+__all__ = ("DetectorTableMeasurementTaskConfig", "DetectorTableMeasurementTask")
+
+
+class DetectorTableMeasurementTaskConnections(MetricConnections,
```
Thanks for pointing out this convention. Done. 

Created separate ticket to do this across faro package:
https://jira.lsstcorp.org/browse/DM-31136

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+            self.prerequisiteInputs.remove("refcat")
+
+
+class DetectorTableMeasurementTaskConfig(CatalogMeasurementBaseTaskConfig,
```
Done.

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+
+            # These default configurations for reference catalogs should go
+            # elsewhere in a common utility.
+            if self.config.connections.refDataset == 'gaia_dr2_20200414':
```
Now putting config overrides in Config class `setDefaults`. Thanks for pointing out this approach, which I had not seen before.

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,129 @@
+            else:
+                inputs['refcat'] = skyCircle.refCat
+
+        outputs = self.run(**inputs)
```
Decided to go with suggestion to pass in variables separately rather than using `self.run(**inputs)`.

bechtol commented:
Thanks for all the feedback. I learned a lot from the comments. I have attempted to implement as many of the suggestions as I could in the recent series of commits. Two questions:

(1) `LoadReferenceCatalogTask` modifies the configuration at this line

https://github.com/lsst/pipe_tasks/blob/f96c35320dc3678d7e47887dbe2ac772fd6200bb/python/lsst/pipe/tasks/loadReferenceCatalog.py#L328

This is now throwing an error 

```
lsst.pex.config.config.FieldValidationError: ConfigurableField 'refObjLoader.refObjLoader' failed validation: Cannot modify a frozen Config
```
This error was not present previously, but appeared when I moved configuration overrides to setDefaults method on the config class. It seems that the configuration is now being frozen. I haven't been able to find a workaround for this issue.

(2) The connections class needs to know the name of the reference catalog to load. Currently, this is determined by the template `{refDataset}`. In addition, `refObjLoader.ref_dataset_name` needs to be set in the configuration of `LoadReferenceCatalogTask`. I think these need to be consistent. Question is whether one of these takes precedent and sets the values of the other (and how to do that correctly), or whether the use should specific both of these in configuration and there should be a check for consistency.

bechtol commented:
Created a separate ticket for adding unit tests for task code.

https://jira.lsstcorp.org/browse/DM-31158

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,144 @@
+    columns = pexConfig.ListField(doc="Columns from sourceTable_visit to load.",
+                                  dtype=str, default=['coord_ra', 'coord_dec', 'detector'])
+
+    refObjLoader = pexConfig.ConfigurableField(
```
Now that I've realized that this is not a `LoadReferenceObjectsTask` class, lets not call it `refObjLoader`, since it behaves rather differently (returning a numpy object). Call it `referenceCatalogLoader` maybe?

bechtol replied:
I like this suggestion. Done.

parejkoj commented:
> LoadReferenceCatalogTask modifies the configuration at line 328

@erykoff : that's not valid Task code, since configs get frozen on task init. You'll want to modify your tests to `freeze()` the config once you've set it up, to better emulate actual Task behavior. (I think there's a way to get this "for free" outside of a pipeline, but I don't recall it right now.)

parejkoj commented on python/lsst/faro/measurement/VisitTableMeasurement.py:
```diff
@@ -0,0 +1,74 @@
+    _DefaultName = "visitTableMeasurementTask"
+
+    def run(self, catalog):
+        return self.measure.run(self.config.connections.metric, catalog)
```
The base class has this as `self.measure.run(cat, self.config.connections.metric)`: what is the correct order? If the base class is correct, you don't need to override `run` here. If the base class is incorrect, why are they different?

bechtol replied:
Now using `run` method inherited from `CatalogMeasurementBaseTask` to enforce consistency. 

```    
    def run(self, catalog, **kwargs):
        return self.measure.run(self.config.connections.metric, catalog, **kwargs)
```

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,144 @@
+
+
+class DetectorTableMeasurementTask(CatalogMeasurementBaseTask):
+    """Base class for science performance metrics measured on single-dector source catalogs."""
```
dector -> detector?

bechtol replied:
Good catch. Done.

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,144 @@
+
+    def run(self, catalog, refcat, refcatCalib):
+        return self.measure.run(self.config.connections.metric, catalog,
+                                refcat=refcat, refcatCalib=refcatCalib)
```
You will need to unify your `run` signatures, since the one you use in your example pipeline above `NumSourcesTask` has quite a different call signature, so it will fail to work here. Is there any `run` args uniformity between the various measurement tasks in faro?

I also see `NumSourcesTask` and `NumSourcesMergeTask` having quite different `run` signatures (and no docstrings!). If those are supposed to be interchangeable, they need to all have the same `run`.

bechtol replied:
Now using the run method inherited from `CatalogMeasurementBaseTask`. Using keyword arguments to handle situation that metrics will require different input arguments.

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,144 @@
+                [butlerQC.quantum.dataId.records['physical_filter'].name],
+                epoch=epoch)
+
+            # Next get unformatted catalog w/ all columns
```
Why do you need both `refcat` and `refcatCalib`? How are they to be used differently, since they nominally contain the same data?

bechtol replied:
The latter object does NOT include all of the columns from the original reference catalog, but rather only the coordinates and colorterm corrected magnitudes. If we want any of the other columns, we need both versions of the catalog.

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,144 @@
+                refCats=inputs.pop('refcat'))
+
+            # Catalog with proper motion and color terms applied
+            refcatCalib = loaderTask.getSkyCircleCatalog(
```
I'm not sure I'd call this `refcatCalib`: refcats are already "calibrated" (they have flux and sky coords, not instFlux and detector positions). It's been proper motion and colorterm corrected, so maybe `refcatUpdated` (I don't really like that either) or something else?

bechtol replied:
Maybe `refCatCorrected`? Note that this object does NOT include all of the columns from the original reference catalog, but rather only the coordinates and colorterm corrected magnitudes.

bechtol replied:
Going with `refCatCorrected`

parejkoj reviewed (changes requested):
Several more comments.

I think the most concerning thing is that there doesn't appear to be any uniformity in the `run` calls, so the faro measurement tasks do not appear to be substitutable with each other (which is the whole point of the design, no?). You need to settle on an API for the various faro components and stick to it.

I'm going to re-echo Simon's call for some testing infrastructure here.

bechtol commented:
The recent commits are intended to start the process of homogenizing the `run` method signatures across `faro`. I'm not sure that it is possible for all the `run` methods to have identical sets of arguments because the different metrics require different information. For example, some metrics require reference catalogs for comparison and others do not. I think at minimum all the measurement tasks associated with source/object catalogs require the metric name and catalog. Perhaps the other arguments would be passed as keywords. 

Doing a quick review of run methods across the different analysis contexts, there some differences, but I think we may want to work through those in a separate PR (see this ticket).
https://jira.lsstcorp.org/browse/DM-31061
Currently, some of the run methods are taking in photometric and astrometric calibrations. Probably we should move those calibration steps out of the task doing the measurement and into the bases classes for the different analysis contexts where application of the calibrations can still be set in configuration. Also, we probably want to pass a filter list instead of dataIds to these measurement tasks. These changes would help to homogenize the `run` method signatures considerably.

Added unit tests for `run` methods of detector and visit measurement tasks. It would be good to expand these tests.

Moved most of the functionality for accessing reference catalogs out of the `DetectorTableMeasurement*` classes and into the base classes for catalog measurements so that can be applied in multiple analysis contexts.

Some clean-up is still needed.

timj commented:
Have you considered having task base classes grouped by the type of metric or required parameters? Then you'd have a way of asking whether two different metric tasks should share a similar `run` method.

bechtol commented:
> Have you considered having task base classes grouped by the type of metric or required parameters? Then you'd have a way of asking whether two different metric tasks should share a similar `run` method.

Currently the base classes are organized by analysis context because these involve different connections, for example, per-detector sources, per-visit sources, catalog of matched sources in a tract (single and multiband), per-tract objects (single and multi-band).  Within a single analysis context, we have some metrics that use external references and some that do not. We are planning to move to using input catalogs in parquet format as much as possible going forward, and hopefully this reduces the need for applying astrometric and photometric calibrations on the fly in the analysis (e.g., separately for each individual detector source catalog, or each individual patch object catalog).

bechtol commented:
Eli made a fix to the `LoadReferenceCatalogTask` so that can handle a frozen config (https://jira.lsstcorp.org/browse/DM-31188, https://github.com/lsst/pipe_tasks/pull/556). With this functionality available, completed the move of reference catalog loading functionality in the `CatalogMeasurementBaseTask` base class.

Here's the passing Jenkins stack-os-matrix Job.

https://ci.lsst.codes/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/34674/pipeline

bechtol commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,144 @@
+    ConfigClass = DetectorTableMeasurementConfig
+    _DefaultName = "detectorTableMeasurementTask"
+
+    def run(self, catalog, refcat, refcatCalib):
```
Now using `run` method from `CatalogMeasurementBaseTask` base class.

parejkoj commented on python/lsst/faro/base/CatalogMeasurementBase.py:
```diff
@@ -16,15 +61,83 @@ class CatalogMeasurementBaseTaskConfig(MetricConfig,
+        skyCircle = loaderTask.refObjLoader.loadSkyCircle(center, radius,
+                                                          loaderTask._referenceFilter,
+                                                          epoch=epoch)
+        if not skyCircle.refCat.isContiguous():
```
This check should not be necessary any more: loadSkyCircle should do this for you.

bechtol replied:
Thanks. This does seem to be the case looking at `loadSkyCircle`. Removed.

parejkoj commented on python/lsst/faro/measurement/DetectorTableMeasurement.py:
```diff
@@ -0,0 +1,98 @@
+        if outputs.measurement is not None:
+            butlerQC.put(outputs, outputRefs)
+        else:
+            self.log.debug("Skipping measurement of {!r} on {} "
```
I'm leaning towards this being an `info` level message, since otherwise things get skipped and you never know about it in normal processing.

bechtol replied:
This relates to another discussion we have been having about how to report various potential failure states. @jeffcarlin may be interested. I suggest we keep this as `debug` level for now, because it is the convention across the `faro` package at the moment, and if we want to make this change, we should make it in multiple places.

parejkoj commented on python/lsst/faro/measurement/VisitTableMeasurement.py:
```diff
@@ -0,0 +1,76 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import lsst.pipe.base as pipeBase
+# from lsst.verify.tasks import MetricConnections
```
Remove unused commented lines, please.

bechtol replied:
done

parejkoj commented on python/lsst/faro/measurement/__init__.py:
```diff
@@ -9,3 +9,7 @@
 from .TractMeasurement import *
 from .PatchMeasurement import *
+
+# Parquet table versions
```
What is this comment for?

bechtol replied:
Currently, we are in the process of transitioning to metric calculations that primarily use parquet table inputs for the source and object catalogs. This is a comment to indicate which of the files are using the parquet table inputs. There is a chance that we might change the directory structure to put the measurements that use parquet table inputs into a separate directory, so my preference is to leave this comment for now.

parejkoj commented on tests/test_tasks.py:
```diff
@@ -0,0 +1,106 @@
+    @classmethod
+    def tearDownClass(cls):
+        '''Delete any variables set in setUpClass.'''
+        del cls.file_map
```
This is unnecessary: builtins get deleted automatically when they go out of scope.

bechtol replied:
Removed

parejkoj commented on tests/test_tasks.py:
```diff
@@ -0,0 +1,106 @@
+        return catalog
+
+    @classmethod
+    def setUpClass(cls):
```
There's nothing in this that requires it be `setUpClass` instead of just `setUp`: the `file_map` is just used as an instance member, not a class member.

bechtol replied:
Thanks. Changed to `setUp`.

parejkoj reviewed (approved):
Please rebase-squash related commits together (for example, you are creating entirely new files here, so all commits that affect one new file should be squashed together. See:

https://developer.lsst.io/work/flow.html#rebase-commits-from-code-reviews-rather-than-having-review-feedback-commits

I've left a few more comments: I'll leave you to sort them out (though I can respond in-threads if we need it).
